7927 unhandled error tb 3

jlitzingerdev opened the pull request:
This introduces a new API to address issues where tracebacks for unhandled errors are not printed.  There has been significant discussion in both the ticket and PRs, so for brevity I'm including links below.

The relevant tickets are:
* https://twistedmatrix.com/trac/ticket/7927
* https://twistedmatrix.com/trac/ticket/9456

This is the result of discussion on a review of an earlier candidate for 7927, located here: https://github.com/twisted/twisted/pull/1022#discussion_r200186589

* [X] There is an associated ticket in Trac. Create a new one at https://twistedmatrix.com/trac/newticket
* [X] The changes pass minimal style checks (see: https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted )
* [X] I have created a newsfragment in src/twisted/newsfragments/ (see: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles )
* [X] I have updated the automated tests.

jlitzingerdev committed:
Introduce a new eventAsText API.

eventAsText facilitates formatting a log event as unicode and
optionally including traceback, system, and timestamp information.
This API should be preferred over the existing formatEvent and
formatEventAsClassicLogText APIs.

One of the problems with the existing APIs is they do not always
include tracebacks, or handle multibyte/non-ascii encodings in the
tracebacks properly.  This API does both.

Update the narrative documentation to recommend this API over the
existing formatEvent API

jlitzingerdev committed:
Use eventAsText in formatEvent and formatEventAsClassicLogText.

The eventAsText API provides all necessary mechanisms to replace the
internals of the existing format APIs without changing the semantics
or form of the output.  This compatibility eliminates the need for the
existing implementations.

jlitzingerdev committed:
Use eventAsText in the default logger to print tracebacks.

Currently, the default logger does not print tracebacks for unhandled
errors.  Using the new eventAsText API will ensure tracebacks are printed.

jlitzingerdev commented:
I'm almost certain the flagged txchecker issue:
`329: W9401: (Used for checking comment format issues.), : Comments should begin with one whitespace` is a bug in twistedchecker.  It is interpreting the `#` in the system tag as a comment.

jlitzingerdev commented:
@glyph FYI, Initial PR for `eventAsText`.

twm commented:
It's definitely a txchecker bug: the same bit me on #1038. https://github.com/twisted/twistedchecker/issues/110

glyph commented on src/twisted/logger/_format.py:
```diff
@@ -287,3 +238,179 @@ def formatWithCall(formatString, mapping):
+    @return: A formatted string with specified options.
+    @rtype: L{unicode}
+
+    @since: Twisted NEXT
```
🎉

glyph commented on src/twisted/logger/_format.py:
```diff
@@ -287,3 +238,179 @@ def formatWithCall(formatString, mapping):
+
+    logText += u"{eventText}".format(eventText=eventText)
+
+    return logText + u"\n"
```
The tacking on of the newline strikes me as a bit odd.  If one is using `eventAsText` directly, one might wish to separate log messages by something other than newlines.  (For example, they might be displayed in some kind of UI widget where the additional newline would space them out too much.)  Why include it?

jlitzingerdev replied:
I think this is actually more an evolutionary artifact of how I developed it.  I'm trying, but can't recall the rationale that led me to leave it in now.  I'll take it out and fix the appropriate places.

glyph commented on src/twisted/logger/_format.py:
```diff
@@ -287,3 +238,179 @@ def formatWithCall(formatString, mapping):
+    if not eventText:
+        return eventText
+
+    logText = u""
```
Normally I wouldn't be so concerned with a tiny micro-optimization, but given that this has historically been a *very* hot code path in Twisted, I think it might be worthwhile to take some trouble to avoid having to rebuild the format string in its entirety on every single call.

(Let's not block on this; it can easily be done in a follow-up PR.)

jlitzingerdev replied:
I'll spend a bit on it, but I won't let it add much delay.

jlitzingerdev replied:
@glyph Since you mentioned this is a hot path, it leads me to another question.  There are helper functions `_formatEvent`, `_formatTraceback` used because (to me) they kept the code more readable.  However, the function call has overhead.  Should those be removed in favor of performance?

glyph commented on docs/core/howto/logger.rst:
```diff
@@ -321,13 +321,13 @@ That interface simply describes a 1-argument callable that takes a ``dict`` , so
     @provider(ILogObserver)
     def simpleObserver(event):
-        print(formatEvent(event))
+        print(eventAsText(event))
```
This example now produces double the newlines.

glyph commented on src/twisted/logger/_format.py:
```diff
@@ -34,27 +34,12 @@ def formatEvent(event):
+        includeTraceback=False,
+        includeTimestamp=False,
+        includeSystem=False,
+    )[:-1]
```
This `:-1` seems like a rather unfortunate consequence of the newline being included.  I realize that this is a "legacy" API now, ideally destined to be deprecated... but still, giving it intimate knowledge of the internals of another function seems weird.  (Why `:-1`?  What about `\r\n` newlines?  What about record separators? etc.)

glyph reviewed (commented):
I like pretty much everything about this except the dang newline.  It just sticks in my craw :).  Hopefully my explanation for why makes sense.  If you feel strongly otherwise please feel free to respond with your reasoning about why you think it should be included (but you probably also want to fix the docs and stuff in that case to not double up on newlines).

Regarding the txchecker failures:

- there are a number of actual lint errors, if you could fix those it would be great
- you can work around the bug by spelling it '\x23' in tests.  I realize this is a bit gross, but the standard thing to do is to work around issues like this and include a link to the txchecker bug so we can clean them up later, rather than littering the codebase with even more warnings that need to be ignored.

I think if you can address this stuff it's basically ready to land.  When you submit for re-review please feel free to be annoying on the mailing list and such so we can do so quickly :).

jlitzingerdev commented:
I'll fix up the style (sorry, that's an annoying miss on my part) and newline issues and resubmit, shouldn't be longer than the upcoming weekend.  I'll also try to get the optimization done.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 7927-unhandled-error-tb-3

jlitzingerdev committed:
Fix style 'W9012:660 Expected 2 blank lines, found 1' error

Methods should have two blank lines between them, not one.

jlitzingerdev committed:
Replace # with \x23 in test files.

This works around a twistedchecker issue:

https://github.com/twisted/twistedchecker/issues/110

Recommended during review.

jlitzingerdev committed:
Remove newline from eventAsText result.

This was raised during review.  There's no value in eventAsText
appending a newline to the resultant text.  The original goal was to
preserve the ending of the existing functions whose internals have
been replaced, but it isn't necessary for it to exist within
eventAsText to achieve that goal.

This change moves the newline out of eventAsText and back into the
appropriate functions (e.g. formatEventAsClassicLogText.

jlitzingerdev committed:
Keep format string static.

Rather than forcing the format string to be rebuilt on each call, have
it static and ensure that the appropriate whitespace/symbols are added
when the components (e.g. system) are set based on the include flags.

Raised during review as this has been hot path code in the past.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 7927-unhandled-error-tb-3

glyph committed:
Merge branch 'trunk' into 7927-unhandled-error-tb-3

hawkowl committed:
Don't print excess newlines in the example

hawkowl commented:
Merged! Thank you.
